### PR TITLE
Add decorations to ir nodes in Painless

### PR DESCRIPTION
**Pull request by jdconrad**
This PR starts to switch all ir nodes to use decorations instead of specific member data. For this specific PR, we add only an expression type decoration to begin with to keep the PR smaller. Other members of the ir nodes will be converted to decorations in a future PR.

The decoration system has two important advantages:
1) The first is it's consistent with the user nodes as generated data is attached to them as a decoration, so we have a clear, consistent model for how to use both trees.
2) This allows additionally generated data to be attached as necessary for optimization phases making the ir tree extendable which is one of our primary, long-term goals.

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java [hunk 1]
```diff
@@ -770,18 +772,22 @@ public void visitBinaryMath(BinaryMathNode irBinaryMathNode, WriteScope writeSco
...
-                methodWriter.writeDynamicBinaryInstruction(irBinaryMathNode.getLocation(), irBinaryMathNode.getExpressionType(),
-                        irLeftNode.getExpressionType(), irRightNode.getExpressionType(), operation, irBinaryMathNode.getFlags());
+                methodWriter.writeDynamicBinaryInstruction(irBinaryMathNode.getLocation(),
+                        irBinaryMathNode.getDecoration(IRDExpressionType.class).getType(),
+                        irLeftNode.getDecoration(IRDExpressionType.class).getType(),
+                        irRightNode.getDecoration(IRDExpressionType.class).getType(),
+                        operation, irBinaryMathNode.getFlags());
             } else {
-                methodWriter.writeBinaryInstruction(irBinaryMathNode.getLocation(), irBinaryMathNode.getExpressionType(), operation);
+                methodWriter.writeBinaryInstruction(
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java:789]: expand this.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java [hunk 2]
```diff
@@ -990,15 +997,16 @@ public void visitInstanceof(InstanceofNode irInstanceofNode, WriteScope writeSco
...
         if (irInstanceofNode.getInstanceType() == def.class) {
             methodWriter.writePop(MethodWriter.getType(expressionType).getSize());
             methodWriter.push(true);
-        } else if (irChildNode.getExpressionType().isPrimitive()) {
+        } else if (irChildNode.getDecoration(IRDExpressionType.class).getType().isPrimitive()) {
             methodWriter.writePop(MethodWriter.getType(expressionType).getSize());
             Class<?> boxedInstanceType = PainlessLookupUtility.typeToBoxedType(instanceType);
-            Class<?> boxedExpressionType = PainlessLookupUtility.typeToBoxedType(irChildNode.getExpressionType());
+            Class<?> boxedExpressionType =
+                    PainlessLookupUtility.typeToBoxedType(irChildNode.getDecoration(IRDExpressionType.class).getType());
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java:1009]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java [hunk 3]
```diff
@@ -1195,11 +1203,11 @@ public void visitTypeCaptureReference(TypedCaptureReferenceNode irTypedCaptureRe
         Variable captured = writeScope.getVariable(irTypedCaptureReferenceNode.getCaptures().get(0));
 
         methodWriter.visitVarInsn(captured.getAsmType().getOpcode(Opcodes.ILOAD), captured.getSlot());
-        Type methodType = Type.getMethodType(
-                MethodWriter.getType(irTypedCaptureReferenceNode.getExpressionType()),
+        Type methodType = Type.getMethodType(MethodWriter.getType(
+                irTypedCaptureReferenceNode.getDecoration(IRDExpressionType.class).getType()),
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java:1207]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java [hunk 4]
```diff
@@ -1195,11 +1203,11 @@ public void visitTypeCaptureReference(TypedCaptureReferenceNode irTypedCaptureRe
...
                 captured.getAsmType());
-        methodWriter.invokeDefCall(irTypedCaptureReferenceNode.getMethodName(),
-                methodType, DefBootstrap.REFERENCE, irTypedCaptureReferenceNode.getExpressionCanonicalTypeName());
+        methodWriter.invokeDefCall(irTypedCaptureReferenceNode.getMethodName(), methodType, DefBootstrap.REFERENCE,
+                irTypedCaptureReferenceNode.getDecoration(IRDExpressionType.class).getCanonicalTypeName());
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java:1210]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 5]
```diff
@@ -973,14 +975,15 @@ public void visitBinary(EBinary userBinaryNode, ScriptScope scriptScope) {
...
+        irExpressionNode.attachDecoration(new IRDExpressionType(valueType));
         scriptScope.putDecoration(userBinaryNode, new IRNodeDecoration(irExpressionNode));
     }
 
     @Override
     public void visitBooleanComp(EBooleanComp userBooleanCompNode, ScriptScope scriptScope) {
         BooleanNode irBooleanNode = new BooleanNode(userBooleanCompNode.getLocation());
-        irBooleanNode.setExpressionType(scriptScope.getDecoration(userBooleanCompNode, ValueType.class).getValueType());
+        irBooleanNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userBooleanCompNode, ValueType.class).getValueType()));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:986]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 6]
```diff
@@ -1008,7 +1011,8 @@ public void visitExplicit(EExplicit userExplicitNode, ScriptScope scriptScope) {
     @Override
     public void visitInstanceof(EInstanceof userInstanceofNode, ScriptScope scriptScope) {
         InstanceofNode irInstanceofNode = new InstanceofNode(userInstanceofNode.getLocation());
-        irInstanceofNode.setExpressionType(scriptScope.getDecoration(userInstanceofNode, ValueType.class).getValueType());
+        irInstanceofNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userInstanceofNode, ValueType.class).getValueType()));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1015]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 7]
```diff
@@ -1091,7 +1098,8 @@ public void visitNewArray(ENewArray userNewArrayNode, ScriptScope scriptScope) {
     public void visitNewObj(ENewObj userNewObjectNode, ScriptScope scriptScope) {
         NewObjectNode irNewObjectNode = new NewObjectNode(userNewObjectNode.getLocation());
 
-        irNewObjectNode.setExpressionType(scriptScope.getDecoration(userNewObjectNode, ValueType.class).getValueType());
+        irNewObjectNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userNewObjectNode, ValueType.class).getValueType()));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1102]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 8]
```diff
@@ -1151,15 +1159,17 @@ public void visitCallLocal(ECallLocal callLocalNode, ScriptScope scriptScope) {
             irInvokeCallMemberNode.addArgumentNode(injectCast(userArgumentNode, scriptScope));
         }
 
-        irInvokeCallMemberNode.setExpressionType(scriptScope.getDecoration(callLocalNode, ValueType.class).getValueType());
+        irInvokeCallMemberNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(callLocalNode, ValueType.class).getValueType()));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1163]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 9]
```diff
@@ -1151,15 +1159,17 @@ public void visitCallLocal(ECallLocal callLocalNode, ScriptScope scriptScope) {
...
 
         scriptScope.putDecoration(callLocalNode, new IRNodeDecoration(irInvokeCallMemberNode));
     }
 
     @Override
     public void visitBooleanConstant(EBooleanConstant userBooleanConstantNode, ScriptScope scriptScope) {
         ConstantNode irConstantNode = new ConstantNode(userBooleanConstantNode.getLocation());
-        irConstantNode.setExpressionType(scriptScope.getDecoration(userBooleanConstantNode, ValueType.class).getValueType());
+        irConstantNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userBooleanConstantNode, ValueType.class).getValueType()));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1172]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 10]
```diff
@@ -1168,7 +1178,8 @@ public void visitBooleanConstant(EBooleanConstant userBooleanConstantNode, Scrip
     @Override
     public void visitNumeric(ENumeric userNumericNode, ScriptScope scriptScope) {
         ConstantNode irConstantNode = new ConstantNode(userNumericNode.getLocation());
-        irConstantNode.setExpressionType(scriptScope.getDecoration(userNumericNode, ValueType.class).getValueType());
+        irConstantNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userNumericNode, ValueType.class).getValueType()));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1182]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 11]
```diff
@@ -1177,7 +1188,8 @@ public void visitNumeric(ENumeric userNumericNode, ScriptScope scriptScope) {
     @Override
     public void visitDecimal(EDecimal userDecimalNode, ScriptScope scriptScope) {
         ConstantNode irConstantNode = new ConstantNode(userDecimalNode.getLocation());
-        irConstantNode.setExpressionType(scriptScope.getDecoration(userDecimalNode, ValueType.class).getValueType());
+        irConstantNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userDecimalNode, ValueType.class).getValueType()));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1192]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 12]
```diff
@@ -1335,7 +1347,8 @@ public void visitFunctionRef(EFunctionRef userFunctionRefNode, ScriptScope scrip
             irReferenceNode = typedInterfaceReferenceNode;
         }
 
-        irReferenceNode.setExpressionType(scriptScope.getDecoration(userFunctionRefNode, ValueType.class).getValueType());
+        irReferenceNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userFunctionRefNode, ValueType.class).getValueType()));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1351]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 13]
```diff
@@ -1393,7 +1406,8 @@ public void visitNewArrayFunctionRef(ENewArrayFunctionRef userNewArrayFunctionRe
 
         irClassNode.addFunctionNode(irFunctionNode);
 
-        irReferenceNode.setExpressionType(scriptScope.getDecoration(userNewArrayFunctionRefNode, ValueType.class).getValueType());
+        irReferenceNode.attachDecoration(
+                new IRDExpressionType(scriptScope.getDecoration(userNewArrayFunctionRefNode, ValueType.class).getValueType()));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1410]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 14]
```diff
@@ -1665,20 +1679,21 @@ public void visitBrace(EBrace userBraceNode, ScriptScope scriptScope) {
 
             if (write == false || compound) {
                 LoadMapShortcutNode irLoadMapShortcutNode = new LoadMapShortcutNode(location);
-                irLoadMapShortcutNode.setExpressionType(scriptScope.getDecoration(userBraceNode, ValueType.class).getValueType());
+                irLoadMapShortcutNode.attachDecoration(
+                        new IRDExpressionType(scriptScope.getDecoration(userBraceNode, ValueType.class).getValueType()));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1683]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 15]
```diff
@@ -1687,7 +1702,8 @@ public void visitBrace(EBrace userBraceNode, ScriptScope scriptScope) {
 
             if (write == false || compound) {
                 LoadListShortcutNode irLoadListShortcutNode = new LoadListShortcutNode(location);
-                irLoadListShortcutNode.setExpressionType(scriptScope.getDecoration(userBraceNode, ValueType.class).getValueType());
+                irLoadListShortcutNode.attachDecoration(
+                        new IRDExpressionType(scriptScope.getDecoration(userBraceNode, ValueType.class).getValueType()));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1706]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 16]
```diff
@@ -1715,7 +1731,8 @@ public void visitCall(ECall userCallNode, ScriptScope scriptScope) {
                 irCallSubDefNode.addArgumentNode((ExpressionNode)visit(userArgumentNode, scriptScope));
             }
 
-            irCallSubDefNode.setExpressionType(scriptScope.getDecoration(userCallNode, ValueType.class).getValueType());
+            irCallSubDefNode.attachDecoration(
+                    new IRDExpressionType(scriptScope.getDecoration(userCallNode, ValueType.class).getValueType()));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1735]: pull into intermediate variable.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 17]
```diff
@@ -1751,7 +1768,8 @@ public void visitCall(ECall userCallNode, ScriptScope scriptScope) {
                 irInvokeCallNode.addArgumentNode(injectCast(userCallArgumentNode, scriptScope));
             }
 
-            irInvokeCallNode.setExpressionType(scriptScope.getDecoration(userCallNode, ValueType.class).getValueType());;
+            irInvokeCallNode.attachDecoration(
+                    new IRDExpressionType(scriptScope.getDecoration(userCallNode, ValueType.class).getValueType()));
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1772]: pull into intermediate variable.

**Review by stu-elastic — APPROVED** [review]
Create an issue to be able to print out the tree, with decorations, in every phase.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java:789]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java:1009]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java:1207]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java (hunk 4 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java:1210]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 5 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:986]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 6 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1015]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 7 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1102]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 8 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1163]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 9 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1172]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 10 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1182]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 11 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1192]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 12 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1351]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 13 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1410]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 14 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1683]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 16 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1735]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 15 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1706]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 17 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:1772]: Done.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review! Updated each place you mentioned in the PR to make more local vars for readability. Added (https://github.com/elastic/elasticsearch/issues/64076) for printing trees. Will commit as soon as CI passes.

**Pull request merged**